core/src/test/java/io/aiven/kafka/tieredstorage/fetch/index/MemorySegmentIndexesCacheTest.java

```diff
@@ -0,0 +1,291 @@
+            );
+            assertThat(offsetIndex).hasBinaryContent(OFFSET_INDEX);
+            assertThat(cache.cache.asMap()).hasSize(1);
+            assertThat(cache.get(
+                key,
+                IndexType.OFFSET,
+                offsetIndexSupplier
+            )).hasBinaryContent(OFFSET_INDEX);
+            // Fetching chunk 0 multiple times from the cache to guarantee that during the next fetch of not-yet-cached
+            // chunk chunk 0 will not be evicted as least frequently accessed.
+            for (int i = 0; i < 50; i++) {
+                assertThat(cache.get(
+                    key,
+                    IndexType.OFFSET,
+                    offsetIndexSupplier
+                )).hasBinaryContent(OFFSET_INDEX);
+            }
+            verifyNoMoreInteractions(offsetIndexSupplier);
+            assertThat(cache.cache.asMap()).hasSize(1);
+
+            final InputStream timeIndex = cache.get(
+                key,
+                IndexType.TIMESTAMP,
+                timeIndexSupplier
+            );
+            assertThat(timeIndex).hasBinaryContent(TIME_INDEX);
+            assertThat(cache.cache.asMap().size()).isGreaterThanOrEqualTo(1);
+
+            await()
+                .atMost(Duration.ofSeconds(30)) // increase to reduce chance of flakiness
+                .pollDelay(Duration.ofSeconds(2))
+                .pollInterval(Duration.ofMillis(10))
+                .until(() -> !mockingDetails(removalListener).getInvocations().isEmpty());
+
+            assertThat(cache.cache.asMap()).hasSize(1);
+            verify(removalListener).onRemoval(any(SegmentIndexKey.class), any(), eq(RemovalCause.SIZE));
+        }
+    }
+
+    @Nested
+    class ErrorHandlingTests {
+        private static final String TEST_EXCEPTION_MESSAGE = "test_message";
+
+        private final Map<String, String> configs = Map.of(
+            "retention.ms", "-1",
+            "size", "-1"
+        );
+
+        @BeforeEach
+        void setUp() {
+            cache.configure(configs);
+        }
+
+        @Test
+        void failedFetching() {
+            when(offsetIndexSupplier.get())
+                .thenThrow(new RuntimeException(new StorageBackendException(TEST_EXCEPTION_MESSAGE)));
+            when(timeIndexSupplier.get())
+                .thenThrow(new RuntimeException(new IOException(TEST_EXCEPTION_MESSAGE)));
+
+            final ObjectKey key = objectKeyFactory.key(REMOTE_LOG_SEGMENT_METADATA, ObjectKeyFactory.Suffix.INDEXES);
+            assertThatThrownBy(() -> cache
+                .get(key, IndexType.OFFSET, offsetIndexSupplier))
+                .isInstanceOf(StorageBackendException.class)
+                .hasMessage(TEST_EXCEPTION_MESSAGE);
+            assertThatThrownBy(() -> cache
+                .get(key, IndexType.TIMESTAMP, timeIndexSupplier))
+                .isInstanceOf(IOException.class)
+                .hasMessage(TEST_EXCEPTION_MESSAGE);
+        }
```

Comment:
[Nit] This probably asks for splitting into two tests

Reply:
Sure. Turning this into a parametrized test.